app/models/product_for_cart.rb

```diff
@@ -0,0 +1,118 @@
+class ProductForCart
+
+  include TextHelpers::Translation
+
+  attr_accessor :error_message
+  attr_accessor :error_path
+
+  def initialize(product)
+    @product = product
+  end
+
+  def purchasable_by?(acting_user, session_user)
+    error = checks(acting_user, session_user).detect do |check|
+      result = check.call
+      [error_path, error_message].any?(&:present?) || result == false
+    end
+    error.nil?
+  end
+
+  protected
+
+  def translation_scope
+    "models.#{self.class.name.underscore}"
+  end
+
+  private
+
+  attr_accessor :product
+
+  def checks(acting_user, session_user)
+    [
+      product_is_accessible?(session_user),
```

Comment:
I like how this setup reads, but the fact that each method needs to return a proc/lambda feels a little funky, but it's not ultimately bad.

Reply:
What if we did something like:

```ruby
[
  [:rpoduct_is_accessible?, session_user],
  [:user_is_present?, acting_user],
  ...
]
```

Then the loop could possibly be:
```ruby
checks(acting_user, session_user).all? do |check, *args|
  method(check).call(*args)
end
```

(I haven't actually verified this, especially with the splats). But that could clean up the need for the procs. Just an idea.

Reply:
I had that before, but I ended up preferring the procs over like using `method` to extract a callable, and coming up with serialization of the name and arguments as an array.

Are you okay with keeping the procs? I think it’s more straightforward, but yeah — the fact that we want to stop after the first error we encounter makes this not ideal.